Compiler/BackEnd/BackendDAEOptimize.mo

```diff
@@ -110,6 +110,122 @@ algorithm
   outDAE.shared := shared;
 end simplifyAllExpressions;
 
+// =============================================================================
+// simplifyInStream
+//
+// OM introduces $OMC$PosetiveMax which can simplified using min or max attribute
+// see Modelica spec for inStream
+// author: Vitalij Ruge
+// see. #3885, 4441
+// =============================================================================
+
+public function simplifyInStream
+  input output BackendDAE.BackendDAE dae;
+protected
+  BackendDAE.Shared shared = dae.shared;
+  BackendDAE.EqSystems eqs = dae.eqs;
+  list<BackendDAE.Variables> vars = list(eq.orderedVars for eq in eqs);
+algorithm
+  // from the description inputs are part of globalKnownVars and localKnownVars :(
+  vars := shared.globalKnownVars :: vars; // need inputs with min = 0 or max = 0
+  vars := shared.localKnownVars :: vars;
+  _ := BackendDAEUtil.traverseBackendDAEExpsNoCopyWithUpdate(dae, simplifyInStreamWork, vars);
+end simplifyInStream;
+
+
+protected function simplifyInStreamWork
+  input DAE.Exp inExp;
+  input list<BackendDAE.Variables> inVars;
+  output DAE.Exp outExp;
+  output list<BackendDAE.Variables> outVars = inVars;
+algorithm
+  outExp := Expression.traverseExpBottomUp(inExp, simplifyInStreamWork2, outVars);
+end simplifyInStreamWork;
+
+protected function simplifyInStreamWork2
+  input DAE.Exp inExp;
+  input list<BackendDAE.Variables> inVars;
+  output DAE.Exp outExp;
+  output list<BackendDAE.Variables> outVars = inVars;
+algorithm
+  outExp := match(inExp)
+              local DAE.Type tp;
+              DAE.ComponentRef cr;
+              DAE.Exp e, expr;
+              Option<DAE.Exp> eMin;
+              Option<DAE.Exp> eMax;
+              Boolean isZero;
+
+
+            case DAE.CALL(path=Absyn.IDENT("$OMC$PositiveMax"),expLst={e as DAE.CREF(componentRef=cr), expr})
+              algorithm
+               (eMin, eMax) := simplifyInStreamWorkExpresion(cr, outVars);
+               isZero := simplifyInStreamWorkSimplify(eMin, false);
+               tp := ComponentReference.crefTypeFull(cr);
+               then
+                  if isZero then e else Expression.makePureBuiltinCall("max", {e, expr}, tp);
+
+            case DAE.CALL(path=Absyn.IDENT("$OMC$PositiveMax"),expLst={e as DAE.UNARY(DAE.UMINUS(tp), DAE.CREF(componentRef=cr)), expr})
+              algorithm
+               (eMin, eMax) := simplifyInStreamWorkExpresion(cr, outVars);
+               isZero := simplifyInStreamWorkSimplify(eMax, true);
+               then
+                  if isZero then Expression.createZeroExpression(tp) else Expression.makePureBuiltinCall("max", {e, expr}, tp);
+
+            case DAE.CALL(path=Absyn.IDENT("$OMC$PositiveMax"),expLst={e, expr})
+               guard Expression.isZero(e)
+               then e;
```

Comment:
This seems totally wrong. If e==0 you should probably return eps, right?

Reply:
Unclear:
 1)  `0.max=0.min = 0` or `0.nominal = 0 //  eps = nominal?`  
 2) https://trac.openmodelica.org/OpenModelica/ticket/4441 is valid 
   - accept the effect of elimination  

if 1) and 2) are correct then 0 is correct. Otherwise I would accept eps, too

Reply:
The 3.4 specification does not use the min or max attributes of the expression anywhere; only the actual value of the expression. And since a tool is free to merge min/max attributes (which is especially done for aliases), you should just assume for a literal value, if it is within the min/max bounds you can tighten the bounds anyway.

Reply:
In this case  (eMin, eMax) defined as (SOME(e), SOME(e)) and we are in the same cases as about, right?

Reply:
the case 0/0 is missing, maybe we can keep eps ... for the moment

Reply:
if it's induce some bugs